rclpy/rclpy/client.py

```diff
@@ -81,20 +81,21 @@ def call(self, request: SrvTypeRequest) -> SrvTypeResponse:
         if not isinstance(request, self.srv_type.Request):
             raise TypeError()
 
-        event = threading.Event()
-
-        def unblock(future):
-            nonlocal event
-            event.set()
-
         future = self.call_async(request)
-        future.add_done_callback(unblock)
 
-        # Check future.done() before waiting on the event.
-        # The callback might have been added after the future is completed,
-        # resulting in the event never being set.
-        if not future.done():
-            event.wait()
+        while not future.done():
+            with self.handle:
+                header, response = self.__client.take_response(self.srv_type.Response)
```

Comment:
just comments, i did not run this code for test.

- if response is not ready for the request, this `while loop` will consume cpu resource until the response is received?
- if another thread requests with the same `Client` before, it might take out the unrelated response and return?

i may be missing something though.

Reply:
> if response is not ready for the request, this `while loop` will consume cpu resource until the response is received?

Yes. This part I could use some help on. How does executor avoid polling clients which do not have responses ready?

> if another thread requests with the same `Client` before, it might take out the unrelated response and return?

I don't think this should be a problem. The `Client` processes responses the same way that the executor does so even if it processes an unrelated response it should be fine.

Reply:
It looks like this tries to get the client response while the executor is blocked. It will still deadlock if the service server is in the same executor, or if the service servier relies on information from something in the same executor before responding.